zio-parser/shared/src/main/scala/zio/parser/Parser.scala

```diff
@@ -1439,3 +1439,76 @@ object Parser {
     final case class AllBranchesFailed[Err](left: ParserError[Err], right: ParserError[Err]) extends ParserError[Err]
   }
 }
+
+final case class ParseFailure[+Err](input: String, error: ParserError[Err]) {
+  def pretty: String = {
+    val sb = new StringBuilder
+
+    def errorBuilder(error: ParserError[Err]): Unit =
+      error match {
+        case ParserError.Failure(nameStack, position, failure) =>
+          sb.append(s"Failure at position $position: $failure")
+          if (nameStack.nonEmpty) {
+            sb.append(" in ")
+            sb.append(nameStack.mkString(" -> "))
+          }
+        case ParserError.UnknownFailure(nameStack, position)   =>
+          sb.append(s"Unknown failure at position $position")
+          if (nameStack.nonEmpty) {
+            sb.append(" in ")
+            sb.append(nameStack.mkString(" -> "))
+          }
+        case ParserError.UnexpectedEndOfInput(nameStack)       =>
+          sb.append("Unexpected end of input")
+          if (nameStack.nonEmpty) {
+            sb.append(" in ")
+            sb.append(nameStack.mkString(" -> "))
+          }
+        case ParserError.NotConsumedAll(nameStack, position)   =>
+          sb.append(s"Parser did not consume all input at position $position")
+          if (nameStack.nonEmpty) {
+            sb.append(" in ")
+            sb.append(nameStack.mkString(" -> "))
+          }
+        case ParserError.AllBranchesFailed(left, right)        =>
+          sb.append("All branches failed: ")
+          errorBuilder(left)
+          sb.append(" and ")
+          errorBuilder(right)
+      }
+
+    def printError(pos: Int) = {
+      val lines          = input.split("\n")
+      val positions      = lines.scanLeft(0)(_ + _.length + 1) // +1 for newline character
+      val errorLineIndex = positions.indexWhere(_ > pos) - 1
+
+      val contextRange = (math.max(errorLineIndex - 2, 0), math.min(errorLineIndex + 2, lines.length - 1))
```

Comment:
For multiline input, output the previous two lines and the next two lines, replace the rest with an ellipsis